### PR TITLE
Sample locked override

### LINKED ISSUE #238: Allow admin users to edit details of locked samples
From: Daniel McDonald
Date: Monday, June 29, 2020 at 9:54 AM
To: "Birmingham, Amanda"
Cc: Dom Nguyen, Daniel Joseph Hakim, Diaz Pardo, Edgar
Subject: Re: Samples to plate

:+1:

On Jun 29, 2020, at 9:54 AM, Birmingham, Amanda wrote:

Daniel, the private api prevents changing details on a sample that has already been scanned.  I suggest that we extend the api functionality so that it allows such changes ONLY if the user has an admin token.  Sound ok, or is that still too much risk?
Amanda
 
From: Dom Nguyen
Date: Monday, June 29, 2020 at 9:12 AM
To: Daniel McDonald
Cc: Daniel Joseph Hakim, "Birmingham, Amanda", "Diaz Pardo, Edgar"
Subject: Re: Samples to plate
 
Yup!  This has to do with the issue where samples were accidentally scanned in/listed as "sample-is-valid" preemptively, (ex. sample was a "left hand" type, is actually blood, but listed as "valid" & have a scan in date in Flight).  Thus, I can't edit it using my admin tools, otherwise I receive this error:
 
Error:

{"code":422,"message":"Sample edits locked: Sample already received"}

Best,
Dominic
 
 
On Mon, Jun 29, 2020 at 9:06 AM Daniel McDonald wrote:
Dom, can you expand on the not editable comment?
 
On Mon, Jun 29, 2020, 09:03 Dom Nguyen wrote:
Important to note that there are a handful of samples on that list that have been mis-assigned, based on what the wet lab has suggested (and confirmed by the participants), but I was not able to update this information since they had accidentally been scanned in and are not editable. 
 
I also did a quick spot check pulling random barcodes from the "Samples to be Scanned In" tab that were previously not consented/assigned to a sample type. There are some samples that have recently been updated and are good to be processed, but are not on this list (X00167937, 157616, 157617, 157618, 149225).

Best, 
Dominic

### PR DESCRIPTION
**Pull request by wasade**
Fixes #238 

This pull request adds in the ability to override modification of a locked sample. No API changes were necessary: the override is implicit for administrative users. As a caveat, this assumes the administrator knows what they are doing which is reasonable given the powers they wield in other areas. 

One awkward item though is that this PR does introduce the concept of admin into the primary API module.

### CODE REVIEW

**Comment by dhakim87** [comment]
"One awkward item though is that this PR does introduce the concept of admin into the primary API module."

I don't think that's awkward at all.  This type of change makes total sense to me, as long as we're able to make the policies on which roles can do what clear and centralized.  I'm quite happy with this change.

**Comment by wasade** [comment]
@dhakim87 excellent then :)

File: microsetta_private_api/api/tests/test_api.py [hunk 1]
```diff
@@ -1300,6 +1350,129 @@ def test_associate_answered_survey_to_sample_success(self):
...
+
+        # verify the sample was changed and visible from the participant
+        get_resp = self.client.get(
+            '%s?%s' % (sample_url, self.default_lang_querystring),
+            headers=self.dummy_auth
+        )
+        self.assertEqual(200, get_resp.status_code)
+        self.assertEqual(get_resp.json['sample_site'], 'Saliva')
+
+        # delete as admin so that tearDown doesn't require admin
```
Comment (wasade) [inline microsetta_private_api/api/tests/test_api.py:1468]: Update `tearDown` to use the overrides

Comment (wasade) [reply microsetta_private_api/api/tests/test_api.py]: Fixed in 62c8586

**Comment by wasade** [comment]
@AmandaBirmingham or @dhakim87, is this good to merge?

**Comment by wasade** [comment]
Actually, sorry, we had discussed it this past week and was good to merge following the override piece going into teardown.

**Pull request merged**